shared/src/balancer_event_handler.rs

```diff
@@ -3,195 +3,183 @@ use crate::{
-    MinimalSwapInfo = 1,
-    /// only allows two tokens to be registered. This achieves the lowest possible swap gas cost.
-    /// Like minimal swap info Pools, these are called via IMinimalSwapInfoPool.
-    TwoToken = 2,
-}
-
-impl PoolSpecialization {
-    fn new(specialization: u8) -> Result<Self> {
-        match specialization {
-            0 => Ok(Self::General),
-            1 => Ok(Self::MinimalSwapInfo),
-            2 => Ok(Self::TwoToken),
-            t => Err(anyhow!("Invalid PoolSpecialization value {} (> 2)", t)),
-        }
-    }
+    /// Temporary storage for WeightedPools having errored on data fetching
+    pending_pools: HashSet<WeightedPoolBuilder>,
+    #[derivative(Debug = "ignore")]
+    data_fetcher: Box<dyn PoolDataFetching>,
 }
 
 impl BalancerPools {
-    fn try_upgrade(&mut self) {
-        for (pool_id, pool_builder) in self.pending_pools.clone() {
-            if let Some(weighted_pool) = pool_builder.into_pool() {
-                // When upgradable, delete pending pool and add to valid pools
-                tracing::info!("Upgrading Pool Builder with id {:?}", pool_id);
-                self.pools.insert(pool_id, weighted_pool.clone());
-                self.pending_pools.remove(&pool_id);
-                for token in weighted_pool.tokens {
-                    self.pools_by_token
-                        .entry(token)
-                        .or_default()
-                        .insert(pool_id);
-                }
+    async fn try_upgrade(&mut self) -> Result<()> {
+        for pool_builder in self.pending_pools.clone() {
+            let weighted_pool = pool_builder.into_pool(&*self.data_fetcher).await?;
+            let pool_id = weighted_pool.pool_id;
+            // delete pending pool and add to valid pools
+            tracing::info!("Upgrading Pool Builder with id {:?}", pool_id);
+            self.pools.insert(pool_id, weighted_pool.clone());
+            self.pending_pools.remove(&pool_builder);
+            for token in weighted_pool.tokens {
+                self.pools_by_token
+                    .entry(token)
+                    .or_default()
+                    .insert(pool_id);
             }
         }
+        Ok(())
     }
 
-    // All insertions happen in one transaction.
-    fn insert_events(&mut self, events: Vec<(EventIndex, BalancerEvent)>) -> Result<()> {
-        for (index, event) in events {
-            match event {
-                BalancerEvent::PoolRegistered(event) => self.insert_pool(index, event),
-                BalancerEvent::TokensRegistered(event) => self.insert_token_data(index, event),
-            };
+    async fn insert_events(&mut self, events: Vec<(EventIndex, PoolCreated)>) -> Result<()> {
+        for (index, creation) in events {
+            self.insert_pool(index, creation)
         }
-        // In the future, when processing TokensDeregistered we may have to downgrade the result.
-        self.try_upgrade();
+        self.try_upgrade().await?;
```

Comment:
Why do we still need the try upgrade logic? Can't we now that there is only one event not always immediately create the struct from the event data?

Reply:
The upgrade logic is meant to make the pool data fetching a parallelized EVM call.

Reply:
Ok. Yes we can in fact we can do all this directly inside insert_events instead of needing `try_upgrade`